EIPS/eip-5727.md

```diff
@@ -0,0 +1,596 @@
+     *   MUST revert if the `dates` is in the past.
+     *   MUST revert if the length of `tokenIds` and `dates` do not match.
+     * @param tokenIds The tokens whose expiry dates are set
+     * @param dates The expire dates to set
+     */
+    function setBatchExpiryDates(
+        uint256[] memory tokenIds,
+        uint256[] memory dates
+    ) external;
+}
+```
+
+#### Shadow
+```solidity
+pragma solidity ^0.8.0;
+
+import "./IERC5727.sol";
+
+/**
+ * @title ERC5727 Soulbound Token Shadow Interface
+ * @dev This extension allows restricting the visibility of specific soulbound tokens.
+ */
+interface IERC5727Shadow is IERC5727 {
+    /**
+     * @notice Shadow a token.
+     * @dev MUST revert if the `tokenId` token does not exists.
+     * @param tokenId The token to shadow
+     */
+    function shadow(uint256 tokenId) external;
+
+    /**
+     * @notice Reveal a token.
+     * @dev MUST revert if the `tokenId` token does not exists.
+     * @param tokenId The token to reveal
+     */
+    function reveal(uint256 tokenId) external;
+}
+```
+
+#### SlotEnumerable
+```solidity
+pragma solidity ^0.8.0;
+
+import "./IERC5727.sol";
+import "./IERC5727Enumerable.sol";
+
+/**
+ * @title ERC5727 Soulbound Token Slot Enumerable Interface
+ * @dev This extension allows querying information about slots.
+ */
+interface IERC5727SlotEnumerable is IERC5727, IERC5727Enumerable {
+    /**
+     * @notice Get the total number of slots.
+     * @return The total number of slots.
+     */
+    function slotCount() external view returns (uint256);
+
+    /**
+     * @notice Get the slot with `index` among all the slots.
+     * @dev MUST revert if the `index` exceed the total number of slots.
+     * @param index The index of the slot queried for
+     * @return The slot is queried for
+     */
+    function slotByIndex(uint256 index) external view returns (uint256);
+
+    /**
+     * @notice Get the number of tokens in a slot.
+     * @dev MUST revert if the slot does not exist.
+     * @param slot The slot whose number of tokens is queried for
+     * @return The number of tokens in the `slot`
+     */
+    function tokenSupplyInSlot(uint256 slot) external view returns (uint256);
+
+    /**
+     * @notice Get the tokenId with `index` of the `slot`.
+     * @dev MUST revert if the `index` exceed the number of tokens in the `slot`.
+     * @param slot The slot whose token is queried for.
+     * @param index The index of the token queried for
+     * @return The token is queried for
+     */
+    function tokenInSlotByIndex(uint256 slot, uint256 index)
+        external
+        view
+        returns (uint256);
+}
+```
+
+## Rationale
+### Token storage model
+We adopt semi-fungible token storage models designed to support both fungible and non-fungible tokens, inspired by [EIP-3525](./eip-3525.md). We found that such a model is better suited to the representation of SBT than the model used in [EIP-1155](./eip-1155.md).
```

Comment:
You might not want to directly link to EIP-3525 because this proposal will not be able to advance independently.